File: ViewletCreatorIOS/ViewletCreator/Classes/ViewletCreator.swift
//
//  ViewletCreator.swift
//  Viewlet creator Pod
//
//  Library: viewlet creator
//  The main interface to register and create new viewlets
//

import UIKit

public protocol Viewlet {
    
    func create() -> UIView
    @discardableResult func update(view: UIView, attributes: [String: Any], parent: UIView?, binder: ViewletBinder?) -> Bool
    func canRecycle(view: UIView, attributes: [String: Any]) -> Bool
    
}

public class ViewletCreator {
    
    // ---
    // MARK: Singleton instance
    // ---

    private static let shared = ViewletCreator()
    
    
    // ---
    // MARK: Members
    // ---

    private var registeredViewlets = [String: Viewlet]()
    private var registeredStyles = [String: [String: [String: Any]]]()
    private var mergeSubAttributes = [String]()
    private var excludeAttributes = [String]()
    

    // ---
    // MARK: Initialization
    // ---
    
    private init() {
    }
    
    
    // ---
    // MARK: Registration
    // ---
    
    public static func register(name: String, viewlet: Viewlet) {
        shared.registeredViewlets[name] = viewlet
    }
    
    public static func registerStyle(viewletName: String, styleName: String, styleAttributes: [String: Any]?) {
        var viewletStyle = shared.registeredStyles[viewletName]
        if viewletStyle == nil {
            viewletStyle = [:]
        }
        if styleAttributes == nil {
            let _ = viewletStyle?.removeValue(forKey: styleName)
        } else {
            viewletStyle?[styleName] = styleAttributes
        }
        shared.registeredStyles[viewletName] = viewletStyle!
    }
    
    public static func registeredViewletNames() -> [String] {
        return shared.registeredViewlets.keys.map({ $0 })
    }
    
    public static func setMergeSubAttribbutes(_ attributeNames: [String]) {
        shared.mergeSubAttributes = attributeNames
    }
    
    public static func setExcludeAttributes(_ attributeNames: [String]) {
        shared.excludeAttributes = attributeNames
    }
    

    // ---
    // MARK: Create and update
    // ---

    public static func create(attributes: [String: Any], parent: UIView? = nil, binder: ViewletBinder? = nil) -> UIView? {
        if let viewletName = findViewletNameInAttributes(attributes) {
            if let viewlet = shared.registeredViewlets[viewletName] {
                let view = viewlet.create()
                if let mergedAttributes = processedAttributes(given: attributes, fallback: attributesForStyle(viewletName: viewletName, styleName: ViewletConvUtil.asString(value: attributes["viewletStyle"])), mergeSubAttributes: shared.mergeSubAttributes, excludeAttributes: shared.excludeAttributes) {
                    viewlet.update(view: view, attributes: mergedAttributes, parent: parent, binder: binder)
                }
                return view
            }
        }
        return nil
    }
    
    @discardableResult
    public static func inflateOn(view: UIView, attributes: [String: Any]?, parent: UIView? = nil, binder: ViewletBinder? = nil) -> Bool {
        if attributes == nil {
            return false
        }
        if let viewletName = findViewletNameInAttributes(attributes!) {
            if let viewlet = shared.registeredViewlets[viewletName] {
                if let mergedAttributes = processedAttributes(given: attributes, fallback: attributesForStyle(viewletName: viewletName, styleName: ViewletConvUtil.asString(value: attributes?["viewletStyle"])), mergeSubAttributes: shared.mergeSubAttributes, excludeAttributes: shared.excludeAttributes) {
                    return viewlet.update(view: view, attributes: mergedAttributes, parent: parent, binder: binder)
                }
            }
        }
        return false
    }
    
    public static func canRecycle(view: UIView?, attributes: [String: Any]?) -> Bool {
        if view != nil && attributes != nil {
            if let viewlet = findViewletInAttributes(attributes!) {
                return viewlet.canRecycle(view: view!, attributes: attributes!)
            }
        }
        return false
    }
    
    public static func findViewletInAttributes(_ attributes: [String: Any]) -> Viewlet? {
        if let viewletName = attributes["viewlet"] as? String {
            return shared.registeredViewlets[viewletName]
        }
        return nil
    }
    
    public static func findViewletNameInAttributes(_ attributes: [String: Any]) -> String? {
        return attributes["viewlet"] as? String
    }
    
    
    // ---
    // MARK: Sub-viewlet utilities
    // ---
    
    public static func attributesForSubViewlet(_ subViewletItem: Any?) -> [String: Any]? {
        if let attributes = subViewletItem as? [String: Any] {
            if let viewletName = findViewletNameInAttributes(attributes) {
                return processedAttributes(given: attributes, fallback: attributesForStyle(viewletName: viewletName, styleName: ViewletConvUtil.asString(value: attributes["viewletStyle"])), mergeSubAttributes: shared.mergeSubAttributes, excludeAttributes: shared.excludeAttributes)
            }
        }
        return nil
    }

    public static func attributesForSubViewletList(_ subViewletItemList: Any?) -> [[String: Any]] {
        var viewletItemList: [[String: Any]] = []
        if let itemList = subViewletItemList as? [[String: Any]] {
            for item in itemList {
                if let viewletName = findViewletNameInAttributes(item) {
                    if let resultAttributes = processedAttributes(given: item, fallback: attributesForStyle(viewletName: viewletName, styleName: ViewletConvUtil.asString(value: item["viewletStyle"])), mergeSubAttributes: shared.mergeSubAttributes, excludeAttributes: shared.excludeAttributes) {
                        viewletItemList.append(resultAttributes)
                    }
                }
            }
        }
        return viewletItemList
    }


    // ---
    // MARK: Attribute processing
    // ---
    
    private static func attributesForStyle(viewletName: String, styleName: String?) -> [String: Any]? {
        if let viewletStyles = shared.registeredStyles[viewletName] {
            if styleName == "default" {
                return viewletStyles[styleName!]
            }
            return mergedAttributes(given: viewletStyles[styleName ?? ""], fallback: viewletStyles["default"])
        }
        return nil
    }
    
    private static func processedAttributes(given: [String: Any]?, fallback: [String: Any]?, mergeSubAttributes: [String], excludeAttributes: [String]) -> [String: Any]? {
        if var result = mergedAttributes(given: given, fallback: fallback) {
            for mergeSubAttribute in mergeSubAttributes {
                if let item = result[mergeSubAttribute] as? [String: Any] {
                    result = mergedAttributes(given: item, fallback: result) ?? result
                }
            }
            for excludeAttribute in excludeAttributes {
                result.removeValue(forKey: excludeAttribute)
            }
            return result
        }
        return nil
    }
    
    private static func mergedAttributes(given: [String: Any]?, fallback: [String: Any]?) -> [String: Any]? {
        // Just return one of the attributes if the other is null
        if fallback == nil {
            return given
        } else if given == nil {
            return fallback
        }
        
        // Merge and return without modifying the originals
        var merged: [String: Any] = [:]
        for (key, value) in given! {
            merged[key] = value
        }
        for (key, value) in fallback! {
            if merged[key] == nil {
                merged[key] = value
            }
        }
        return merged
    }

}

File: ViewletCreatorAndroid/ViewletCreatorLib/src/main/java/com/crescentflare/viewletcreator/ViewletCreator.java
package com.crescentflare.viewletcreator;

import android.content.Context;
import android.view.View;
import android.view.ViewGroup;

import com.crescentflare.viewletcreator.binder.ViewletBinder;
import com.crescentflare.viewletcreator.utility.ViewletMapUtil;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Viewlet creator: manages viewlets
 * The main interface to register and create new view components
 */
@SuppressWarnings({"unused", "WeakerAccess"})
public class ViewletCreator
{
    // ---
    // Singleton instance
    // ---

    private static ViewletCreator instance = new ViewletCreator();


    // ---
    // Registered viewlet list
    // ---

    private Map<String, Viewlet> registeredViewlets = new HashMap<>();
    private Map<String, Map<String, Map<String, Object>>> registeredStyles = new HashMap<>();
    private List<String> mergeSubAttributes = new ArrayList<>();
    private List<String> excludeAttributes = new ArrayList<>();


    // ---
    // Initialization
    // ---

    private ViewletCreator()
    {
    }


    // ---
    // Registration
    // ---

    public static void registerViewlet(@Nullable String name, @Nullable Viewlet viewlet)
    {
        if (name != null)
        {
            if (viewlet != null)
            {
                instance.registeredViewlets.put(name, viewlet);
            }
            else
            {
                instance.registeredViewlets.remove(name);
            }
        }
    }

    public static void registerStyle(@Nullable String viewletName, @Nullable String styleName, @Nullable Map<String, Object> styleAttributes)
    {
        if (viewletName != null && styleName != null)
        {
            // First register that viewlet has styles (if not present)
            Map<String, Map<String, Object>> viewletStyles = instance.registeredStyles.get(viewletName);
            if (viewletStyles == null)
            {
                viewletStyles = new HashMap<>();
                instance.registeredStyles.put(viewletName, viewletStyles);
            }

            // Register or remove style
            if (styleAttributes != null)
            {
                viewletStyles.put(styleName, styleAttributes);
            }
            else
            {
                viewletStyles.remove(styleName);
            }
        }
    }

    @NotNull
    public static List<String> registeredViewletNames()
    {
        return new ArrayList<>(instance.registeredViewlets.keySet());
    }

    public static void setMergeSubAttributes(List<String> attributeNames)
    {
        instance.mergeSubAttributes = attributeNames;
    }

    public static void setExcludeAttributes(List<String> attributeNames)
    {
        instance.excludeAttributes = attributeNames;
    }


    // ---
    // Create and update
    // ---

    @Nullable
    public static View create(@NotNull Context context, @Nullable Map<String, Object> attributes, @Nullable ViewGroup parent)
    {
        return create(context, attributes, parent, null);
    }

    @Nullable
    public static View create(@NotNull Context context, @Nullable Map<String, Object> attributes, @Nullable ViewGroup parent, @Nullable ViewletBinder binder)
    {
        String viewletName = findViewletNameInAttributes(attributes);
        if (viewletName != null)
        {
            Viewlet viewlet = instance.registeredViewlets.get(viewletName);
            if (viewlet != null)
            {
                View view = viewlet.create(context);
                Map<String, Object> mergedAttributes = processAttributes(attributes, attributesForStyle(viewletName, ViewletMapUtil.optionalString(attributes, "viewletStyle", null)), instance.mergeSubAttributes, instance.excludeAttributes);
                viewlet.update(view, mergedAttributes, parent, binder);
                return view;
            }
        }
        return null;
    }

    public static boolean inflateOn(@NotNull View view, @Nullable Map<String, Object> attributes, @Nullable ViewGroup parent)
    {
        return inflateOn(view, attributes, parent, null);
    }

    public static boolean inflateOn(@NotNull View view, @Nullable Map<String, Object> attributes, @Nullable ViewGroup parent, @Nullable ViewletBinder binder)
    {
        String viewletName = findViewletNameInAttributes(attributes);
        if (viewletName != null)
        {
            Viewlet viewlet = instance.registeredViewlets.get(viewletName);
            if (viewlet != null)
            {
                Map<String, Object> mergedAttributes = processAttributes(attributes, attributesForStyle(viewletName, ViewletMapUtil.optionalString(attributes, "viewletStyle", null)), instance.mergeSubAttributes, instance.excludeAttributes);
                return viewlet.update(view, mergedAttributes, parent, binder);
            }
        }
        return false;
    }

    public static boolean canRecycle(@Nullable View view, @Nullable Map<String, Object> attributes)
    {
        if (view != null && attributes != null)
        {
            Viewlet viewlet = findViewletInAttributes(attributes);
            if (viewlet != null)
            {
                return viewlet.canRecycle(view, attributes);
            }
        }
        return false;
    }

    @Nullable
    public static Viewlet findViewletInAttributes(@Nullable Map<String, Object> attributes)
    {
        String viewletName = ViewletMapUtil.optionalString(attributes, "viewlet", null);
        if (viewletName != null)
        {
            return instance.registeredViewlets.get(viewletName);
        }
        return null;
    }

    @Nullable
    public static String findViewletNameInAttributes(@Nullable Map<String, Object> attributes)
    {
        return ViewletMapUtil.optionalString(attributes, "viewlet", null);
    }


    // ---
    // Sub-viewlet utilities
    // ---

    @Nullable
    public static Map<String, Object> attributesForSubViewlet(@Nullable Object subViewletItem)
    {
        Map<String, Object> attributes = ViewletMapUtil.asStringObjectMap(subViewletItem);
        if (attributes != null)
        {
            String viewletName = findViewletNameInAttributes(attributes);
            if (viewletName != null)
            {
                return processAttributes(attributes, attributesForStyle(viewletName, ViewletMapUtil.optionalString(attributes, "viewletStyle", null)), instance.mergeSubAttributes, instance.excludeAttributes);
            }
        }
        return null;
    }

    public static List<Map<String, Object>> attributesForSubViewletList(@Nullable Object subViewletItemList)
    {
        List<Map<String, Object>> viewletItemList = new ArrayList<>();
        if (subViewletItemList instanceof List<?>)
        {
            List<?> itemList = (List<?>) subViewletItemList;
            for (Object item : itemList)
            {
                Map<String, Object> attributes = ViewletMapUtil.asStringObjectMap(item);
                if (attributes != null)
                {
                    String viewletName = findViewletNameInAttributes(attributes);
                    if (viewletName != null)
                    {
                        viewletItemList.add(processAttributes(attributes, attributesForStyle(viewletName, ViewletMapUtil.optionalString(attributes, "viewletStyle", null)), instance.mergeSubAttributes, instance.excludeAttributes));
                    }
                }
            }
        }
        return viewletItemList;
    }


    // ---
    // Attribute processing
    // ---

    @Nullable
    private static Map<String, Object> attributesForStyle(@Nullable String viewletName, @Nullable String styleName)
    {
        if (viewletName != null)
        {
            Map<String, Map<String, Object>> viewletStyles = instance.registeredStyles.get(viewletName);
            if (viewletStyles != null)
            {
                if (styleName != null && styleName.equals("default"))
                {
                    return viewletStyles.get(styleName);
                }
                return mergeAttributes(viewletStyles.get(styleName), viewletStyles.get("default"));
            }
        }
        return null;
    }

    @NotNull
    private static Map<String, Object> processAttributes(@Nullable Map<String, Object> givenAttributes, @Nullable Map<String, Object> fallbackAttributes, List<String> mergeSubAttributes, List<String> excludeAttributes)
    {
        Map<String, Object> result = mergeAttributes(givenAttributes, fallbackAttributes);
        for (String mergeSubAttribute : mergeSubAttributes)
        {
            Map<String, Object> item = ViewletMapUtil.asStringObjectMap(result.get(mergeSubAttribute));
            if (item != null)
            {
                result = mergeAttributes(item, result);
            }
        }
        for (String excludeAttribute : excludeAttributes)
        {
            result.remove(excludeAttribute);
        }
        return result;
    }

    @NotNull
    private static Map<String, Object> mergeAttributes(@Nullable Map<String, Object> givenAttributes, @Nullable Map<String, Object> fallbackAttributes)
    {
        // Just return one of the attributes if the other is null
        if (fallbackAttributes == null)
        {
            return givenAttributes != null ? givenAttributes : new HashMap<String, Object>();
        }
        else if (givenAttributes == null)
        {
            return fallbackAttributes;
        }

        // Merge and return without modifying the originals
        Map<String, Object> mergedAttributes = new HashMap<>();
        for (String key : givenAttributes.keySet())
        {
            mergedAttributes.put(key, givenAttributes.get(key));
        }
        for (String key : fallbackAttributes.keySet())
        {
            if (!mergedAttributes.containsKey(key))
            {
                mergedAttributes.put(key, fallbackAttributes.get(key));
            }
        }
        return mergedAttributes;
    }


    // ---
    // The viewlet creation and updating interface
    // ---

    public interface Viewlet
    {
        @NotNull
        View create(@NotNull Context context);

        boolean update(@NotNull View view, @NotNull Map<String, Object> attributes, @Nullable ViewGroup parent, @Nullable ViewletBinder binder);
        boolean canRecycle(@NotNull View view, @NotNull Map<String, Object> attributes);
    }
}
